File: cpp-std/exer02b01-producer-consumer.cpp
/*
THE PRODUCER-CONSUMER PROBLEM

SOLUTION TYPE B: USING SEMAPHORES
    Version B01: 1 slow producer, 1 fast consumer
*/


#include <iostream>
#include <queue>
#include <chrono>
#include <thread>
#include <semaphore>
using namespace std;



using cntsemaphore = std::counting_semaphore<>;



void producer(
    cntsemaphore* semFill,
    cntsemaphore* semEmpty,
    queue<int>* q
) {
    int i = 1;

    for (;; ++i) {
        semEmpty->acquire();

        q->push(i);
        std::this_thread::sleep_for(std::chrono::seconds(1));

        semFill->release();
    }
}



void consumer(
    cntsemaphore* semFill,
    cntsemaphore* semEmpty,
    queue<int>* q
) {
    int data = 0;

    for (;;) {
        semFill->acquire();

        data = q->front();
        q->pop();

        cout << "Consumer " << data << endl;

        semEmpty->release();
    }
}



int main() {
    cntsemaphore semFill(0);   // item produced
    cntsemaphore semEmpty(1);  // remaining space in queue

    queue<int> q;


    auto thProducer = std::thread(producer, &semFill, &semEmpty, &q);
    auto thConsumer = std::thread(consumer, &semFill, &semEmpty, &q);


    thProducer.join();
    thConsumer.join();


    return 0;
}

File: python/exer02b01_producer_consumer.py
'''
THE PRODUCER-CONSUMER PROBLEM

SOLUTION TYPE B: USING SEMAPHORES
    Version B01: 1 slow producer, 1 fast consumer
'''

import time
import threading



def producer(
    sem_fill: threading.Semaphore,
    sem_empty: threading.Semaphore,
    q: list
):
    i = 1

    while True:
        sem_empty.acquire()
        q.append(i)
        time.sleep(1)
        sem_fill.release()
        i += 1



def consumer(
    sem_fill: threading.Semaphore,
    sem_empty: threading.Semaphore,
    q: list
):
    while True:
        sem_fill.acquire()
        data = q.pop(0)
        print('Consumer', data)
        sem_empty.release()



s_fill = threading.Semaphore(0)  # item produced
s_empty = threading.Semaphore(1) # remaining space in queue
que = []

threading.Thread(target=producer, args=(s_fill, s_empty, que)).start()
threading.Thread(target=consumer, args=(s_fill, s_empty, que)).start()
